src/main/kotlin/com/terraformation/backend/nursery/db/BatchStore.kt

```diff
@@ -148,16 +161,104 @@ class BatchStore(
     )
   }
 
-  private fun insertQuantityHistoryRow(row: BatchesRow, historyType: BatchQuantityHistoryType) {
+  fun updateQuantities(
+      batchId: BatchId,
+      version: Int,
+      germinating: Int,
+      notReady: Int,
+      ready: Int,
+      historyType: BatchQuantityHistoryType,
+  ) {
+    if (germinating < 0 || notReady < 0 || ready < 0) {
+      throw IllegalArgumentException("Quantities may not be negative")
+    }
+
+    requirePermissions { updateBatch(batchId) }
+
+    dslContext.transaction { _ ->
+      updateVersionedBatch(batchId, version) { update ->
+        update
+            .set(GERMINATING_QUANTITY, germinating)
+            .set(NOT_READY_QUANTITY, notReady)
+            .set(READY_QUANTITY, ready)
+            .let {
+              if (historyType == BatchQuantityHistoryType.Observed) {
+                it.set(LATEST_OBSERVED_GERMINATING_QUANTITY, germinating)
+                    .set(LATEST_OBSERVED_NOT_READY_QUANTITY, notReady)
+                    .set(LATEST_OBSERVED_READY_QUANTITY, ready)
+                    .set(LATEST_OBSERVED_TIME, clock.instant())
+              } else {
+                it
+              }
+            }
+      }
+
+      insertQuantityHistoryRow(batchId, germinating, notReady, ready, historyType)
+    }
+  }
+
+  /**
+   * Applies an update to a batch if the caller-supplied version number is up to date.
+   *
+   * Automatically updates `modified_by`, `modified_time`, and `version`.
+   *
+   * @param func Function to add `set` statements to an `UPDATE` query. This is called with the
+   * [BATCHES] table as its receiver so that lambda functions can refer to column names without
+   * having to qualify them (that is, `set(FOO, value)` instead of `set(BATCHES.FOO, value)`).
+   * @throws BatchStaleException The batch's version number in the database wasn't the same as
+   * [version].
+   */
+  private fun updateVersionedBatch(
+      batchId: BatchId,
+      version: Int,
+      func: Batches.(UpdateSetFirstStep<BatchesRecord>) -> UpdateSetMoreStep<BatchesRecord>
```

Comment:
So this allows you to inject updates that are namespaced to the `BATCHES` table, right?

Reply:
Kind of. This is arguably me trying to be too clever, but the idea is that if this were just
```
func: (UpdateSetFirstStep<BatchesRecord>) -> UpdateSetMoreStep<BatchesRecord>
```
then the lambda functions would end up needing to do
```
it.set(BATCHES.READY_QUANTITY, readyQuantity)
```
whereas with `Batches` as the receiver type, they can do
```
it.set(READY_QUANTITY, readyQuantity)
```
which is both more concise and also hopefully subtly enforces the notion that the lambda function should only be setting columns on the `batches` table rather than referencing other random tables.